include/utils.hxx

```diff
@@ -166,13 +181,30 @@ public:
 
   Tensor() : n1(0), n2(0), n3(0) {};
   Tensor(size_type n1, size_type n2, size_type n3) : n1(n1), n2(n2), n3(n3) {
-    data = Array<T>(n1*n2*n3);
+    ASSERT2(n1 >= 0);
+    ASSERT2(n2 >= 0);
+    ASSERT2(n3 >= 0);
+    data.resize(n1 * n2 * n3);
   }
   Tensor(const Tensor &other) : n1(other.n1), n2(other.n2), n3(other.n3), data(other.data) {
     // Prevent copy on write for Tensor
     data.ensureUnique();
   }
 
+  /// Resize the Tensor to \p new_size_1 by \p new_size_2 by \p new_size_3
+  ///
+  /// Note that this invalidates the existing data!
+  void resize(size_type new_size_1, size_type new_size_2, size_type new_size_3) {
+    ASSERT2(new_size_1 >= 0);
```

Comment:
I'm not sure what it means for one of `n1`, `n2` and `n3` to be zero here -- clearly the effect is the same as for `Array` and we do explicitly initialise `n*` to 0 in the empty constructor so I guess it just means empty. Should we add a comment/note to explain this.

Reply:
I'll make a note of this in the `Matrix/Tensor` docstring